hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/AssociateTokenRecipientsStep.java

```diff
@@ -54,57 +63,80 @@ public void doIn(@NonNull final TransferContext transferContext) {
+                        receiverId, tokenId, token, accountStore, tokenRelStore, handleContext);
+                if (newAssociation != null) {
+                    newAssociations.add(newAssociation);
+                }
             }
         }
+
+        for (TokenAssociation newAssociation : newAssociations) {
+            recordBuilder.addAutomaticTokenAssociation(newAssociation);
+        }
     }
 
     /**
      * Associates the token with the account if it is not already associated. It is auto-associated only if there are
      * open auto-associations available on the account.
      *
-     * @param accountId     The account to associate the token with
-     * @param tokenId       The tokenID of the token to associate with the account
-     * @param token         The token to associate with the account
-     * @param accountStore  The account store
-     * @param tokenRelStore The token relation store
-     * @param handleContext The context
-     * @param recordBuilder The record builder
+     * @param accountId         The account to associate the token with
+     * @param tokenId           The tokenID of the token to associate with the account
+     * @param token             The token to associate with the account
+     * @param accountStore      The account store
+     * @param tokenRelStore     The token relation store
+     * @param handleContext     The context
      */
-    private void validateAndAutoAssociate(
+    private TokenAssociation validateAndBuildAutoAssociation(
             @NonNull final AccountID accountId,
             @NonNull final TokenID tokenId,
             @NonNull final Token token,
             @NonNull final WritableAccountStore accountStore,
             @NonNull final WritableTokenRelationStore tokenRelStore,
-            @NonNull final HandleContext handleContext,
-            @NonNull final CryptoTransferRecordBuilder recordBuilder) {
+            @NonNull final HandleContext handleContext) {
         final var account = getIfUsable(accountId, accountStore, handleContext.expiryValidator(), INVALID_ACCOUNT_ID);
         final var tokenRel = tokenRelStore.get(accountId, tokenId);
 
         if (tokenRel == null && account.maxAutoAssociations() > 0) {
+            validateFalse(token.hasKycKey(), ACCOUNT_KYC_NOT_GRANTED_FOR_TOKEN);
+            validateFalse(token.accountsFrozenByDefault(), ACCOUNT_FROZEN_FOR_TOKEN);
             final var newRelation = autoAssociate(account, token, accountStore, tokenRelStore, handleContext);
-            recordBuilder.addAutomaticTokenAssociation(
-                    asTokenAssociation(newRelation.tokenId(), newRelation.accountId()));
+            return asTokenAssociation(newRelation.tokenId(), newRelation.accountId());
         } else {
             validateTrue(tokenRel != null, TOKEN_NOT_ASSOCIATED_TO_ACCOUNT);
+            validateFalse(tokenRel.frozen(), ACCOUNT_FROZEN_FOR_TOKEN);
+            return null;
```

Comment:
This is kind of awkward, isn't it? Sorry about that, maybe passing the list in is better. I dunno. We'll fix it in a different PR if we need to.